File: sentinelpass-cli/src/commands/registry.rs
use crate::commands::service_client::{self as sc, Backend};
use crate::RegistryCommands;
use anyhow::Result;
use sentinelpass_core::registry::{Criticality, EntityKind, RegistryOverview, RotationStatus};
use sentinelpass_protocol::service::VaultOp;
use std::path::PathBuf;

pub fn handle_registry_command(
    vault_path: PathBuf,
    command: &crate::RegistryCommands,
) -> Result<()> {
    match command {
        RegistryCommands::EntityAdd {
            name,
            kind,
            criticality,
            notes,
            rotation_interval_days,
        } => handle_entity_add(
            vault_path,
            name,
            kind,
            criticality.as_str(),
            notes.as_deref(),
            *rotation_interval_days,
        ),
        RegistryCommands::EntityList => handle_entity_list(vault_path),
        RegistryCommands::EntityDelete { name } => handle_entity_delete(vault_path, name),
        RegistryCommands::Assign {
            entry_id,
            entity,
            label,
        } => handle_assign(vault_path, *entry_id, entity, label.as_deref()),
        RegistryCommands::MarkRotated { entry_id } => handle_mark_rotated(vault_path, *entry_id),
        RegistryCommands::Unassign { entry_id } => handle_unassign(vault_path, *entry_id),
        RegistryCommands::ExpiresAt {
            entry_id,
            timestamp,
        } => handle_expires_at(vault_path, *entry_id, *timestamp),
        RegistryCommands::Status => handle_status(vault_path),
        RegistryCommands::Report { only_issues } => handle_report(vault_path, *only_issues),
    }
}

fn parse_kind(value: &str) -> Result<EntityKind> {
    EntityKind::parse(value).map_err(|e| anyhow::anyhow!("{}", e))
}

fn parse_criticality(value: &str) -> Result<Criticality> {
    Criticality::parse(value).map_err(|e| anyhow::anyhow!("{}", e))
}

/// Fetch the registry overview through the service backend and decode into
/// the core render type (one conversion point for all registry renders).
fn fetch_overview(backend: &Backend, include_strength: bool) -> Result<RegistryOverview> {
    let value = sc::expect_report(backend.call(VaultOp::RegistryOverview { include_strength })?)?;
    serde_json::from_value(value)
        .map_err(|e| anyhow::anyhow!("failed to decode registry overview: {}", e))
}

fn handle_entity_add(
    vault_path: PathBuf,
    name: &str,
    kind: &str,
    criticality: &str,
    notes: Option<&str>,
    rotation_interval_days: Option<i64>,
) -> Result<()> {
    let kind = parse_kind(kind)?;
    let criticality = parse_criticality(criticality)?;
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    let entity = match backend.call(VaultOp::EntityAdd {
        name: name.trim().to_string(),
        kind: kind.as_str().to_string(),
        criticality: criticality.as_str().to_string(),
        notes: notes
            .map(str::trim)
            .filter(|value| !value.is_empty())
            .map(str::to_string),
        rotation_interval_days,
    })? {
        sentinelpass_protocol::service::VaultOpResult::Entity(entity) => entity,
        other => anyhow::bail!("unexpected response: {other:?}"),
    };
    println!("Entity registered: {} ({})", entity.name, entity.entity_id);
    println!(
        "  kind={} criticality={} rotation-interval={}",
        entity.kind,
        entity.criticality,
        entity
            .rotation_interval_days_override
            .map(|days| format!("{}d (override)", days))
            .unwrap_or_else(|| "policy default".to_string())
    );
    Ok(())
}

fn handle_entity_list(vault_path: PathBuf) -> Result<()> {
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    let overview = fetch_overview(&backend, false)?;

    if overview.entities.is_empty() {
        println!("No entities registered. Use 'sentinelpass registry entity-add' to create one.");
        return Ok(());
    }

    println!(
        "{:<28} {:<16} {:<10} {:<10} {:>10}",
        "NAME", "KIND", "CRITICALITY", "INTERVAL", "CREDENTIALS"
    );
    for summary in &overview.entities {
        let interval = summary
            .entity
            .rotation_interval_days_override
            .map(|days| format!("{}d*", days))
            .unwrap_or_else(|| "default".to_string());
        println!(
            "{:<28} {:<16} {:<10} {:<10} {:>10}",
            summary.entity.name,
            summary.entity.kind.as_str(),
            summary.entity.criticality.as_str(),
            interval,
            summary.credential_count
        );
    }
    println!("\n(* = entity-level override; policy defaults otherwise)");
    Ok(())
}

fn handle_entity_delete(vault_path: PathBuf, name: &str) -> Result<()> {
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    backend.call(VaultOp::EntityDelete {
        name: name.trim().to_string(),
    })?;
    println!("Entity deleted: {}", name.trim());
    Ok(())
}

fn handle_assign(
    vault_path: PathBuf,
    entry_id: i64,
    entity_name: &str,
    label: Option<&str>,
) -> Result<()> {
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    backend.call(VaultOp::EntryAssign {
        entry_id,
        entity: entity_name.trim().to_string(),
        label: label.map(str::to_string),
    })?;
    println!("Entry {} assigned to {}", entry_id, entity_name.trim());
    Ok(())
}

fn handle_mark_rotated(vault_path: PathBuf, entry_id: i64) -> Result<()> {
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    backend.call(VaultOp::EntryMarkRotated { entry_id })?;
    println!(
        "Entry {} marked as rotated (password_rotated_at = now)",
        entry_id
    );
    Ok(())
}

fn handle_unassign(vault_path: PathBuf, entry_id: i64) -> Result<()> {
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    backend.call(VaultOp::EntryUnassign { entry_id })?;
    println!("Entry {} unassigned from its entity", entry_id);
    Ok(())
}

fn handle_expires_at(vault_path: PathBuf, entry_id: i64, timestamp: Option<i64>) -> Result<()> {
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    backend.call(VaultOp::EntrySetExpiresAt {
        entry_id,
        expires_at: timestamp,
    })?;
    match timestamp {
        Some(ts) => println!("Entry {} expiry set to unix timestamp {}", entry_id, ts),
        None => println!("Entry {} expiry cleared", entry_id),
    }
    Ok(())
}

fn handle_status(vault_path: PathBuf) -> Result<()> {
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    let overview = fetch_overview(&backend, false)?;

    let findings = overview
        .posture
        .iter()
        .filter(|entry| entry.status != RotationStatus::Ok)
        .count();

    println!("Registry posture");
    println!(
        "  entities: {} | entries assigned: {} | unassigned: {}",
        overview.entities.len(),
        overview
            .entities
            .iter()
            .map(|summary| summary.credential_count)
            .sum::<i64>(),
        overview.unassigned_entries
    );
    println!("  reuse clusters: {}", overview.reuse_clusters.len());
    println!("  rotation findings: {}", findings);

    if !overview.reuse_clusters.is_empty() {
        println!("\nReuse clusters (same secret across entries):");
        for cluster in &overview.reuse_clusters {
            println!(
                "  [{}] {}",
                cluster.size,
                cluster
                    .entry_ids
                    .iter()
                    .zip(cluster.titles.iter())
                    .map(|(id, title)| format!("#{} \"{}\"", id, title))
                    .collect::<Vec<_>>()
                    .join(", ")
            );
        }
    }

    print_findings(&overview.posture, false);
    Ok(())
}

fn handle_report(vault_path: PathBuf, only_issues: bool) -> Result<()> {
    let backend = sc::connect(&vault_path, || crate::prompt_master_password(false))?;
    let overview = fetch_overview(&backend, true)?;

    println!(
        "Registry report ({} entries, strength analysis included)",
        overview.posture.len()
    );
    print_findings(&overview.posture, only_issues);
    Ok(())
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn status_label(status: RotationStatus) -> &'static str {
    match status {
        RotationStatus::Ok => "OK     ",
        RotationStatus::DueSoon => "DUE    ",
        RotationStatus::Weak => "WEAK   ",
        RotationStatus::Reused => "REUSED ",
        RotationStatus::Overdue => "OVERDUE",
    }
}

fn print_findings(posture: &[sentinelpass_core::EntryPosture], only_issues: bool) {
    let rows: Vec<_> = posture
        .iter()
        .filter(|entry| !only_issues || entry.status != RotationStatus::Ok)
        .collect();

    if rows.is_empty() {
        println!("\nNo findings.");
        return;
    }

    println!();
    for entry in rows {
        println!(
            "[{}] #{} \"{}\"{}",
            status_label(entry.status),
            entry.entry_id,
            entry.title,
            entry
                .entity_name
                .as_deref()
                .map(|name| format!(" (entity: {})", name))
                .unwrap_or_default()
        );
        println!(
            "         interval={}d age={}d{}{}",
            entry.resolved_interval_days,
            entry.days_since_rotation.unwrap_or(0),
            if entry.tool_managed {
                " tool-managed"
            } else {
                ""
            },
            if let Some(expires_at) = entry.expires_at {
                format!(" expires={}", expires_at)
            } else {
                String::new()
            }
        );
        for reason in &entry.reasons {
            println!("         - {}", reason);
        }
    }
}

File: sentinelpass-ui/registry.ts
/**
 * SentinelPass Desktop UI — credential registry posture dashboard (ADR-001 P2).
 *
 * Read-only: view rotation posture, reuse clusters, and registered entities.
 * All mutation (entity creation, assignment, mark-rotated, expiry) remains a
 * CLI-only workflow for this slice — see ADR-001's migration path.
 */

import { invoke, currentEntry, noSelection, entryDetail } from './state.js';
import { showToast, escapeHtml } from './utils.js';
import { loadEntry } from './entries.js';

// ---------------------------------------------------------------------------
// Types mirroring the Rust structs (sentinelpass-core/src/registry/mod.rs,
// policy.rs). Snake_case fields verbatim to match serde output — no
// camelCase translation, consistent with how entries.ts consumes Entry.
// ---------------------------------------------------------------------------

export type RotationStatus = 'ok' | 'due_soon' | 'weak' | 'reused' | 'overdue';

export type Entity = {
    entity_id: string;
    name: string;
    kind: string;
    criticality: string;
    notes: string | null;
    rotation_interval_days_override: number | null;
    created_at: number;
    modified_at: number;
};

export type EntitySummary = {
    entity: Entity;
    credential_count: number;
};

export type ReuseCluster = {
    size: number;
    entry_ids: number[];
    titles: string[];
};

export type EntryPosture = {
    entry_id: number;
    title: string;
    entity_name: string | null;
    status: RotationStatus;
    reasons: string[];
    resolved_interval_days: number;
    days_since_rotation: number | null;
    reuse_count: number;
    strength_score: number | null;
    tool_managed: boolean;
    expires_at: number | null;
};

export type RegistryOverview = {
    entities: EntitySummary[];
    reuse_clusters: ReuseCluster[];
    posture: EntryPosture[];
    unassigned_entries: number;
};

// ---------------------------------------------------------------------------
// Module state (local — nothing else needs registry data)
// ---------------------------------------------------------------------------

let overview: RegistryOverview | null = null;

const STATUS_LABELS: Record<RotationStatus, string> = {
    ok: 'OK',
    due_soon: 'DUE',
    weak: 'WEAK',
    reused: 'REUSED',
    overdue: 'OVERDUE',
};

// ---------------------------------------------------------------------------
// Data loading
// ---------------------------------------------------------------------------

/**
 * Fetch a cheap (no-decrypt) registry overview on unlock, purely to
 * populate the header badge count. Safe to call even if the registry
 * index hasn't been built yet — the backend sweeps it lazily on demand.
 */
export async function loadRegistryBadgeCount() {
    try {
        const result: RegistryOverview = await invoke('get_registry_overview', {
            includeStrength: false,
        });
        updateRegistryBadge(result.posture.length);
    } catch (error) {
        // Best-effort — a badge-count failure shouldn't surface an error
        // toast on every unlock; the dashboard itself will report errors
        // when actually opened.
        console.warn('Failed to load registry badge count:', error);
    }
}

function updateRegistryBadge(count: number) {
    const badge = document.getElementById('registry-badge');
    if (!badge) {
        return;
    }
    if (count > 0) {
        badge.textContent = String(count);
        badge.classList.remove('hidden');
    } else {
        badge.classList.add('hidden');
    }
}

/**
 * Fetch the full (strength-inclusive) registry overview and render the
 * dashboard. Shows a loading state while in flight — this may include a
 * bounded full-vault decrypt sweep on first open after upgrade.
 */
export async function loadRegistryOverview() {
    const loading = document.getElementById('registry-loading');
    loading?.classList.remove('hidden');
    try {
        overview = await invoke('get_registry_overview', { includeStrength: true });
        renderRegistryDashboard(overview);
        updateRegistryBadge(overview.posture.length);
    } catch (error) {
        showToast(error, 'error');
    } finally {
        loading?.classList.add('hidden');
    }
}

// ---------------------------------------------------------------------------
// Rendering
// ---------------------------------------------------------------------------

function renderRegistryDashboard(data: RegistryOverview) {
    const onboardingBanner = document.getElementById('registry-onboarding-banner');
    if (onboardingBanner) {
        onboardingBanner.classList.toggle('hidden', data.entities.length > 0);
    }

    renderStatusStrip(data.posture);
    renderFindingsList(data.posture);
    renderEntitiesList(data.entities);
}

function renderStatusStrip(posture: EntryPosture[]) {
    const headline = document.getElementById('registry-status-headline');
    if (!headline) {
        return;
    }
    const findingCount = posture.length;
    headline.textContent =
        findingCount === 0
            ? 'All clear'
            : `${findingCount} ${findingCount === 1 ? 'finding needs' : 'findings need'} attention`;
}

function renderFindingsList(posture: EntryPosture[]) {
    const container = document.getElementById('registry-findings');
    if (!container) {
        return;
    }

    if (posture.length === 0) {
        container.innerHTML = '<div class="registry-list-empty">No rotation findings — everything looks OK.</div>';
        return;
    }

    container.innerHTML = posture
        .map(
            (entry) => `
        <div class="registry-finding-row" data-entry-id="${entry.entry_id}">
            <div class="registry-finding-row-header" data-action="open-entry">
                <div class="registry-finding-title">${escapeHtml(entry.title)}${
                entry.entity_name
                    ? ` <span class="metadata-muted">&middot; ${escapeHtml(entry.entity_name)}</span>`
                    : ''
            }</div>
                <span class="status-badge status-${entry.status.replace(/_/g, '-')}">${
                STATUS_LABELS[entry.status]
            }</span>
            </div>
            <div class="registry-finding-reasons">
                interval=${entry.resolved_interval_days}d
                ${entry.days_since_rotation !== null ? ` age=${entry.days_since_rotation}d` : ''}
                ${entry.tool_managed ? ' &middot; tool-managed' : ''}
                ${entry.reasons.map((reason) => `<br>&bull; ${escapeHtml(reason)}`).join('')}
            </div>
            ${
                entry.status === 'reused'
                    ? `<button class="registry-disclosure-toggle" data-action="toggle-cluster" data-entry-id="${entry.entry_id}">shared with ${
                          entry.reuse_count - 1
                      } other${entry.reuse_count - 1 === 1 ? '' : 's'} &rarr;</button>
                       <div class="registry-cluster-members hidden" data-cluster-for="${entry.entry_id}"></div>`
                    : ''
            }
        </div>
    `
        )
        .join('');

    container.querySelectorAll('[data-action="open-entry"]').forEach((el) => {
        el.addEventListener('click', () => {
            const entryId = parseInt((el.closest('.registry-finding-row') as HTMLElement).dataset.entryId!);
            openEntityFromPosture(entryId);
        });
    });

    container.querySelectorAll('[data-action="toggle-cluster"]').forEach((el) => {
        el.addEventListener('click', (event) => {
            event.stopPropagation();
            const entryId = parseInt((el as HTMLElement).dataset.entryId!);
            toggleReuseDisclosure(entryId);
        });
    });
}

function toggleReuseDisclosure(entryId: number) {
    if (!overview) {
        return;
    }
    const membersContainer = document.querySelector(
        `[data-cluster-for="${entryId}"]`
    ) as HTMLElement | null;
    if (!membersContainer) {
        return;
    }

    if (!membersContainer.classList.contains('hidden')) {
        membersContainer.classList.add('hidden');
        return;
    }

    const cluster = overview.reuse_clusters.find((c) => c.entry_ids.includes(entryId));
    if (!cluster) {
        return;
    }
    const otherTitles = cluster.entry_ids
        .map((id, i) => ({ id, title: cluster.titles[i] }))
        .filter((member) => member.id !== entryId && member.title !== undefined)
        .map((member) => escapeHtml(member.title));
    membersContainer.textContent = otherTitles.length > 0 ? `Also used by: ${otherTitles.join(', ')}` : 'Also used by another entry.';
    membersContainer.classList.remove('hidden');
}

async function openEntityFromPosture(entryId: number) {
    await loadEntry(entryId);
    document.getElementById('registry-dashboard')?.classList.add('hidden');
}

function renderEntitiesList(entities: EntitySummary[]) {
    const container = document.getElementById('registry-entities');
    if (!container) {
        return;
    }

    if (entities.length === 0) {
        container.innerHTML = '';
        return;
    }

    container.innerHTML = entities
        .map(
            (summary) => `
        <div class="registry-entity-row">
            <div class="registry-entity-row-header">
                <div class="registry-finding-title">${escapeHtml(summary.entity.name)}</div>
                <span class="entry-type-pill">${escapeHtml(summary.entity.kind)}</span>
            </div>
            <div class="registry-finding-reasons">
                criticality=${escapeHtml(summary.entity.criticality)} &middot; ${summary.credential_count} credential${
                summary.credential_count === 1 ? '' : 's'
            }
            </div>
        </div>
    `
        )
        .join('');
}

// ---------------------------------------------------------------------------
// Panel visibility
// ---------------------------------------------------------------------------

export function showRegistryDashboard() {
    noSelection?.classList.add('hidden');
    entryDetail?.classList.add('hidden');
    document.getElementById('registry-dashboard')?.classList.remove('hidden');
    void loadRegistryOverview();
}

export function hideRegistryDashboard() {
    document.getElementById('registry-dashboard')?.classList.add('hidden');
    if (currentEntry) {
        entryDetail?.classList.remove('hidden');
    } else {
        noSelection?.classList.remove('hidden');
    }
}
